controllers/metal3.io/hostfirmwarecomponents_controller.go

```diff
@@ -0,0 +1,332 @@
+			reason = reasonInvalidComponent
+			if setUpdatesCondition(generation, &newStatus, info, metal3api.HostFirmwareComponentsValid, metav1.ConditionFalse, reason, fmt.Sprintf("Invalid Firmware Components: %s", err)) {
+				dirty = true
+			}
+		} else if setUpdatesCondition(generation, &newStatus, info, metal3api.HostFirmwareComponentsValid, metav1.ConditionTrue, reason, "") {
+			dirty = true
+		}
+	} else {
+		if setUpdatesCondition(generation, &newStatus, info, metal3api.HostFirmwareComponentsValid, metav1.ConditionTrue, reason, "") {
+			dirty = true
+		}
+		if setUpdatesCondition(generation, &newStatus, info, metal3api.HostFirmwareComponentsChangeDetected, metav1.ConditionFalse, reason, "") {
+			dirty = true
+		}
+	}
+
+	// Update Status if has changed
+	if dirty {
+		info.log.Info("Status for HostFirmwareComponents changed")
+		info.hfc.Status = *newStatus.DeepCopy()
+
+		t := metav1.Now()
+		info.hfc.Status.LastUpdated = &t
+		return newStatus, r.Status().Update(info.ctx, info.hfc)
+	}
+	return newStatus, nil
+}
+
+// Update the HostFirmwareComponents resource using the components from provisioner.
+func (r *HostFirmwareComponentsReconciler) updateHostFirmwareComponents(newStatus metal3api.HostFirmwareComponentsStatus, components []metal3api.FirmwareComponentStatus, info *rhfcInfo) (err error) {
+	dirty := false
+	// change the Components in Status
+	newStatus.Components = components
+	// Check if the components information we retrieved is different from the one in Status
+	componentsInfoMismatch := !reflect.DeepEqual(components, info.hfc.Status.Components)
+	reason := reasonValidComponent
+	generation := info.hfc.GetGeneration()
+	// Log the components we have
+	info.log.Info("firmware components for node", "components", components, "bmh", info.bmh.Name)
+	if componentsInfoMismatch {
+		setUpdatesCondition(generation, &newStatus, info, metal3api.HostFirmwareComponentsChangeDetected, metav1.ConditionTrue, reason, "")
+		dirty = true
+	}
+	if dirty {
+		info.log.Info("Components Status for HostFirmwareComponents changed")
+		info.hfc.Status = *newStatus.DeepCopy()
+
+		t := metav1.Now()
+		info.hfc.Status.LastUpdated = &t
+		return r.Status().Update(info.ctx, info.hfc)
+	}
+	return nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *HostFirmwareComponentsReconciler) SetupWithManager(mgr ctrl.Manager, maxConcurrentReconcile int) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&metal3api.HostFirmwareComponents{}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: maxConcurrentReconcile}).
+		WithEventFilter(
+			predicate.Funcs{
+				UpdateFunc: r.updateEventHandler,
+			}).
+		Complete(r)
+}
+
+func (r *HostFirmwareComponentsReconciler) updateEventHandler(e event.UpdateEvent) bool {
+	r.Log.Info("hostfirmwarecomponents in event handler")
+
+	if e.ObjectNew.GetGeneration() != e.ObjectOld.GetGeneration() {
+		r.Log.Info("returning true as generation changed from event handler")
```

Comment:
nit: Has this message been copied from somewhere? It's not overly helpful for operators.

Reply:
yes, it was from the HostFirmwareSettings, since I was following the pattern we had there. should I just remove the Log or you have other idea?

Reply:
Possibly remove. We're not even returning true here despite what is says. And maybe remove the whole call since it does not do much.